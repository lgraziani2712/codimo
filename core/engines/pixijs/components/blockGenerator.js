/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */
import { Container, Sprite, Texture } from 'pixi.js';

import { ZERO, ONE } from 'core/constants/numbers';

import { type CodimoComponent } from './componentGenerator';

const SHADOW_WALL = 0.075;

export const BLOCK_DEFINITIONS = {
  BORDER: 0x7d7d7d,
  PATH: 0x2a2a2a,
  WALL: 0xda4b4b,
};

export type ActivePathBorders = {
  bottom?: boolean,
  left?: boolean,
  right?: boolean,
  top?: boolean,

  [any]: empty,
};
export type Direction = 'top' | 'right' | 'bottom' | 'left';
type CornerDirection = 'topLeft' | 'topRight' | 'bottomLeft' | 'bottomRight';

const cornerGenerator = (position: CornerDirection, size: number, margin: number) => {
  const corner = new Sprite(Texture.WHITE);

  corner.width = corner.height = margin;
  corner.x = position === 'topRight' || position === 'bottomRight' ? size + margin : ZERO;
  corner.y = position === 'bottomLeft' || position === 'bottomRight' ? size + margin : ZERO;

  return corner;
};
const wallGenerator = (
  position: Direction,
  activePathBorders: ActivePathBorders,
  size: number,
  margin: number,
) => {
  const wall = new Sprite(Texture.WHITE);

  // If is the wall is an active path, then it is transparent.
  wall.alpha = activePathBorders[position] ? SHADOW_WALL : ONE;

  wall.width = position === 'right' || position === 'left' ? margin : size;
  wall.height = position === 'right' || position === 'left' ? size : margin;

  return wall;
};
const blockBuilder = (tint: $Values<typeof BLOCK_DEFINITIONS>, size: number, margin: number) => {
  const block = new Sprite(Texture.WHITE);

  block.alpha = BLOCK_DEFINITIONS.PATH === tint ? ZERO : ONE;
  block.tint = tint;
  block.height = block.width = size;
  block.x = block.y = margin;

  return block;
};

/**
 * This visual component represent a block.
 * One step for a movable component means moving from
 * one block to another.
 *
 * @version 1.0.0
 * @param {number} tint Block's colour.
 * @param {number} size Block's size.
 * @param {number} margin Block's margin.
 * @return {CodimoComponent} A visual component representing a block.
 */
const blockGenerator = (
  tint: number,
  size: number,
  margin: number,
) => (x: number, y: number, activePathBorders?: ActivePathBorders = {}): CodimoComponent => {
  const view = new Container();
  const block = blockBuilder(tint, size, margin);
  //////////////////////////////
  // Borders
  //////////////////////////////
  const topWall = wallGenerator('top', activePathBorders, size, margin);
  const rightWall = wallGenerator('right', activePathBorders, size, margin);
  const bottomWall = wallGenerator('bottom', activePathBorders, size, margin);
  const leftWall = wallGenerator('left', activePathBorders, size, margin);
  //////////////////////////////
  // Corners
  //////////////////////////////
  const topLeftCorner = cornerGenerator('topLeft', size, margin);
  const topRightCorner = cornerGenerator('topRight', size, margin);
  const bottomLeftCorner = cornerGenerator('bottomLeft', size, margin);
  const bottomRightCorner = cornerGenerator('bottomRight', size, margin);

  // TODO move it to the generator.
  topWall.x = leftWall.y = bottomWall.x = rightWall.y = margin;
  bottomWall.y = rightWall.x = size + margin;

  view.x = x;
  view.y = y;

  view.addChild(
    block,
    topWall,
    rightWall,
    bottomWall,
    leftWall,
    topLeftCorner,
    topRightCorner,
    bottomLeftCorner,
    bottomRightCorner,
  );

  return {
    view,
  };
};

export default blockGenerator;
