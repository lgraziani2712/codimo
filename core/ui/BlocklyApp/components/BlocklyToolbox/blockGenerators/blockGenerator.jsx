/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */
import React from 'react';

import { type BlocklyBlockValue } from './valueGenerator';
import { type BlocklyBlockField } from './fieldGenerator';

export type BlocklyBlock = {|
  define: 'block',
  type: string,
  props?: Array<BlocklyBlockValue | BlocklyBlockField>,
|};
const blockGenerator = (element: BlocklyBlock, idx: number) => (
  <block key={idx} type={element.type} />
);

export default blockGenerator;
