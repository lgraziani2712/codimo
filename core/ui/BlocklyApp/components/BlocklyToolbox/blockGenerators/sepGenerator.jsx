/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */
import React from 'react';

const DEFAULT_GAAP = 24;

export type BlocklySep = {|
  define: 'sep',
  /**
   * By default every block is separated from its lower neighbour by 24 pixels.
   * This separation may be changed using the 'gap' attribute,
   * which will replace the default gap.
   *
   * @see https://developers.google.com/blockly/guides/configure/web/toolbox#separators
   */
  gaap?: number,
|};
const sepGenerator = (element: BlocklySep, idx: number) => (
  <sep key={idx} gap={element.gaap === undefined ? DEFAULT_GAAP : element.gaap} />
);

export default sepGenerator;
