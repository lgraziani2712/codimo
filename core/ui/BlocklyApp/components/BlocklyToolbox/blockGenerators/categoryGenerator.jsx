/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */
import React from 'react';

import blockGenerator, { type BlocklyBlock } from './blockGenerator';
import sepGenerator, { type BlocklySep } from './sepGenerator';

export type BlocklyCategory = {|
  blocks: Array<BlocklyBlock | BlocklySep>,
  define: 'category',
  name: string,
  color?: number | string,
|};
const categoryGenerator = (element: BlocklyCategory, idx: number) => (
  <category key={idx} name={element.name} color={element.color}>
    {element.blocks.map((block, idx) => (
      block.define === 'block'
        ? blockGenerator(block, idx)
        : sepGenerator(block, idx)
    ))}
  </category>
);

export default categoryGenerator;
