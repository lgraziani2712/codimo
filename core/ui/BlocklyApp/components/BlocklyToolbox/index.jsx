/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */
import React from 'react';

import blockGenerator, { type BlocklyBlock } from './blockGenerators/blockGenerator';
import sepGenerator, { type BlocklySep } from './blockGenerators/sepGenerator';
import categoryGenerator, { type BlocklyCategory } from './blockGenerators/categoryGenerator';

type BlocklyBaseElement = BlocklyBlock | BlocklySep;
export type BlocklyToolboxElement = BlocklyCategory | BlocklyBaseElement;

type Props = {|
  elements: Array<BlocklyToolboxElement>,
  handleWorkspaceCreation: (toolbox: HTMLElement) => void,
|};
/**
 * The BlocklyToolbox container is responsible to parse
 * the activity configuration for Blocky and generate
 * a correct XML representation which will be required
 * for the instanciation of the Blockly workspace.
 *
 * @todo Add example
 * @version 1.0.0
 */
export default class BlocklyToolbox extends React.Component {
  props: Props;
  toolbox: HTMLElement;

  componentDidMount() {
    // https://developers.google.com/blockly/guides/configure/web/toolbox
    this.props.handleWorkspaceCreation(this.toolbox);
  }
  render() {
    const { elements } = this.props;

    return (
      <xml ref={(toolbox: HTMLElement) => this.toolbox = toolbox} style={{ display: 'none' }}>
        {elements.map((element, idx) => {
          if (element.define === 'block') {
            return blockGenerator(element, idx);
          }
          if (element.define === 'sep') {
            return sepGenerator(element, idx);
          }
          if (element.define === 'category') {
            return categoryGenerator(element, idx);
          }
          throw new Error(`The element definition '${element.define}' does not exist`);
        })}
      </xml>
    );
  }
}

// const blocklyGenerators = {
//   block: blockGenerator,
//   category: categoryGenerator,
//   sep: sepGenerator,
// };
//
//
// const generator = blocklyGenerators[element.define];
//
// if (typeof generator !== 'function') {
//   throw new Error(`The element definition '${element.define}' does not exist`);
// }
//
// return generator(element, idx);
