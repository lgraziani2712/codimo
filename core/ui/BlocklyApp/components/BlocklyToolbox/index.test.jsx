/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 * @jest-environment node
 */
/* eslint-disable no-magic-numbers */
import React from 'react';
import renderer from 'react-test-renderer';

import blockGenerator from './blockGenerators/blockGenerator';
import categoryGenerator from './blockGenerators/categoryGenerator';
import sepGenerator from './blockGenerators/sepGenerator';

import BlocklyToolbox from '.';

describe('BlocklyToolbox', () => {
  it('renders correctly an empty toolbox', () => {
    const tree = renderer.create(
      <BlocklyToolbox elements={[]} handleWorkspaceCreation={() => {}} />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
  describe('Generators', () => {
    describe('blockGenerator', () => {
      it('renders correctly without children', () => {
        const description = {
          define: 'block',
          type: 'test_block',
        };
        const tree = renderer.create(
          blockGenerator(description, 0),
        ).toJSON();

        expect(tree).toMatchSnapshot();
      });
      xit('renders correctly with children');
    });
  });
  describe('categoryGenerator', () => {
    it('renders correctly without children', () => {
      const description = {
        define: 'category',
        name: 'test_category',
        color: '#deedee',
        blocks: [],
      };
      const tree = renderer.create(
        categoryGenerator(description, 2),
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });
    it('renders correctly with children', () => {
      const description = {
        define: 'category',
        name: 'test_category',
        color: '#deedee',
        blocks: [{
          define: 'block',
          type: 'test_block',
        }, {
          define: 'sep',
        }],
      };
      const tree = renderer.create(
        categoryGenerator(description, 2),
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
  describe('sepGenerator', () => {
    it('renders correctly', () => {
      const description = {
        define: 'sep',
        gaap: 32,
      };
      const tree = renderer.create(
        sepGenerator(description, 3),
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
