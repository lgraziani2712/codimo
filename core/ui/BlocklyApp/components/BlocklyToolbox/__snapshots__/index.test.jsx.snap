// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlocklyToolbox Generators blockGenerator renders correctly without children 1`] = `
<block
  type="test_block"
/>
`;

exports[`BlocklyToolbox categoryGenerator renders correctly with children 1`] = `
<category
  color="#deedee"
  name="test_category"
>
  <block
    type="test_block"
  />
  <sep
    gap={24}
  />
</category>
`;

exports[`BlocklyToolbox categoryGenerator renders correctly without children 1`] = `
<category
  color="#deedee"
  name="test_category"
/>
`;

exports[`BlocklyToolbox renders correctly an empty toolbox 1`] = `
<xml
  style={
    Object {
      "display": "none",
    }
  }
/>
`;

exports[`BlocklyToolbox sepGenerator renders correctly 1`] = `
<sep
  gap={32}
/>
`;
